src/include/storage/index/bplustree.h

```diff
@@ -747,6 +749,501 @@ class BPlusTree : public BPlusTreeBase {
+      } else {
+        current_node = node->GetLowKeyPair().second;
+      }
+
+      // Get the shared latch for next node and release the parent
+      current_node->GetNodeSharedLatch();
+      parent->ReleaseNodeSharedLatch();
+    }
+
+    return current_node;
+  }
+
+  /**
+   * This function returns the last leaf node in the B+ Tree (node that contains
+   * the highest key in the tree)
+   * @return Pointer to the last leaf node
+   */
+  BaseNode *FindLastLeafNode() {
+    root_latch_.LockShared();
+
+    if (root_ == nullptr) {
+      root_latch_.UnlockShared();
+      return nullptr;
+    }
+    BaseNode *current_node = root_;
+    BaseNode *parent = nullptr;
+    // Acquire latch on the root, release root_latch
+    current_node->GetNodeSharedLatch();
+    root_latch_.UnlockShared();
+    // Traversing Down to the correct leaf node
+    while (current_node->GetType() != NodeType::LeafType) {
+      // Set parent for releasing the lock
+      parent = current_node;
+      current_node = current_node->GetHighKeyPair().second;
+      // Get shared latch on current node, and release the parent.
+      current_node->GetNodeSharedLatch();
+      parent->ReleaseNodeSharedLatch();
+    }
+    return current_node;
+  }
+
+  /**
+   * This class implements a bi-directional iterator for the B+ Tree. In order to
+   * fetch an element to start iteration, Begin() function can be used. Then, ++ or --
+   * operators can be used to iterate till End() or REnd() is reached respectively.
+   *
+   * NOTE: In order to avoid holding locks in case of error, user of the iterator should
+   * call Done() to release any lock that is being currently held.
+   */
+  class BPlusTreeIterator {
+    /** Enum that represents the state of the iterator */
+    enum IteratorState { VALID, END, REND, RETRY, INVALID };
+
+    /** Node in which iterator is operating currently */
+    ElasticNode<KeyValuePair> *curr_node_;
+    /** Leaf Node element at which iterator is operating currently */
+    KeyValuePair *curr_key_;
+    /** Internal iterator of the value list on which iterator is operating currently */
+    typename ValueList::iterator curr_val_;
+    /** State of the iterator */
+    IteratorState state_;
+
+    /**
+     * Resets the iterator to represent a dummy iterator.
+     */
+    void ResetIterator() {
+      curr_node_ = nullptr;
+      curr_key_ = nullptr;
+    }
+
+    /**
+     * Sets the state of the iterator to END iterator
+     */
+    void SetEndIterator() {
+      ResetIterator();
+      state_ = END;
+    }
+
+    /**
+     * Sets the state of the iterator to REND iterator
+     */
+    void SetREndIterator() {
+      ResetIterator();
+      state_ = REND;
+    }
+
+    /**
+     * Sets the state of the iterator to RETRY iterator
+     */
+    void SetRetryIterator() {
+      ResetIterator();
+      state_ = RETRY;
+    }
+
+   public:
+    /** Key at the current position of the iterator */
+    KeyType first_;
```

Comment:
Does this mean we do deep copies of current key and value into the iterator? Is there a way we can avoid that?

Reply:
Yeah, did this to make it work similar to a normal iterator. An alternative would be to not expose `first` and `second` as members, but have two public functions `first()` and `second()` or `key()` and `value()`. In this implementation, it will be dereference and copy on every call. 
I can switch to this implementation, technically it should decrease one copy. Currently for every key, we copy it to `first_` and for every value, we copy it to `second_` whenever the iterator moves.  Then there will be another copy to transfer this into `results` vector during a scan. If we switch to having a function instead, the first copy can be avoided.
Do you have any preference on function names? `first()`, `second()` vs `key()`, `value()`.